### PR TITLE
Feat/rule based intent transition

### PR DESCRIPTION
**Pull request by ltbringer**
We require custom logic to handle arbitrary transitions. Giving a detailed look, the transitions appear to have simple relationships that can even be serialized and expressed as configs.

Taking an example:

If we have discovered in an utterance, that the current `state` is `S1` and the intent is `I2` and in this case, we should transition to `I2`.
Then this can be expressed by the following rule:

```python
rules = [{
   rename: "I2",
   depends_on: {
      "intent": "I1",
      "state": "S1"
   }
}]
```

We can do more; say we have discovered 
- entity types as: `["e1", "e2"]` 
- `state` as `s1` 
- and we need to transition to `I2`?

```python
rules = [{
   rename: "I2",
   depends_on: {
      "entities": {
         "intersects": ["e1"]
      },
      "state": "S1"
   }
}]
```

We have also identified that these transitions currently exist for the following properties:
1. The predicted intent with most confidence.
2. The current state
3. The list of entity types
4. The previous intent

By accommodating 5 operations for each of the 4 properties we can produce a moderately strong rule transition, namely:

1. eq - equality check, also the default operation
2. ne - not equal
3. in - containership
4. nin - inverse containership
5. intersects - an intersection between two sets (or array-like) quantities.

This can help in cases where there is a dire shortage of data.

### CODE REVIEW

File: dialogy/plugins/text/output_mods/rule_based_intent_swap.py [hunk 1]
```diff
@@ -8,6 +8,60 @@
 
 
 class RuleBasedIntentSwap(Plugin):
+    """
+    We require custom logic to handle arbitrary transitions. Giving a detailed look, the transitions
+    appear to have simple relationships that can even be serialized and expressed as configs.
+
+    Taking an example:
+
+    If we have discovered in an utterance, that the current state is :code:`S1` and the intent is :code:`I2` and in this case,
```
Comment (biswaroop1547) [inline dialogy/plugins/text/output_mods/rule_based_intent_swap.py:19]: > and the intent is `I2`

it should be - and the intent is `I1` ?

**Review by biswaroop1547 — APPROVED** [review]
lgtm!

**Comment by us241098** [comment]
intent transition depends on entity values too

**Comment by nvinayvarma189** [comment]
is something like this possible?

```
if original_intent == X:
    if previous_intent in some_intent_list:
        original_intent = previous_intent
```
This type of logic is mostly needed for swapping `date_entity_only` intent

**Comment by ltbringer** [comment]
> is something like this possible?
> 
> ```
> if original_intent == X:
>     if previous_intent in some_intent_list:
>         original_intent = previous_intent
> ```
> 
> This type of logic is mostly needed for swapping `date_entity_only` intent

@nvinayvarma189 wouldn't that be already covered by the rule:

```
rules = [{
  rename: "<previous_intent>",
  depends_on: {
     previous_intent: {
        "in": ["..."]
     }
  }
}]
```

**Comment by us241098** [comment]
@ltbringer case like this:

```
if ent.value=="english":
    intents[0].name = "inform_english"
else:
    intents[0].name = "inform_hindi"
```

**Pull request merged**